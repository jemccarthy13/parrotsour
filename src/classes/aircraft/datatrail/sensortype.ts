export enum SensorType {
  RAW,
  ARROW,
}
