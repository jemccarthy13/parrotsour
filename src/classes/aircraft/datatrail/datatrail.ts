import { Point } from "../../point"
import { IDMatrix } from "../id"

// TODO -- DATATRAIL -- register? so that Aircraft can generate one of each type of datatrail?
// TODO -- DATATRAIL -- Support a toggle for the different Sensor Types.....
export abstract class DataTrail {
  private startPos: Point

  constructor(startPos: Point) {
    this.startPos = startPos
  }

  public getStartPos(): Point {
    return this.startPos
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public move(_heading: number): void {
    //console.warn("Moving DataTrail in base class. Will have no effect.")
  }

  abstract getCenterOfMass(heading: number): Point

  abstract draw(heading: number, id: IDMatrix): void
}
