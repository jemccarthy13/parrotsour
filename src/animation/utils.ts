import { PaintBrush } from "../canvas/draw/paintbrush"
import { SensorType } from "../classes/aircraft/datatrail/sensortype"
import { AircraftGroup } from "../classes/groups/group"

/**
 * Internal function to see if a group is near the canvas boundary.
 *
 * Used to change group logic when approaching the edge of the canvas
 * @param group The group to check against boundaries
 * @param dataStyle The current DataTrail type
 */
export function isNearBounds(
  group: AircraftGroup,
  dataStyle: SensorType
): boolean {
  const buffer = 40
  const sX = group.getCenterOfMass(dataStyle).x
  const sY = group.getCenterOfMass(dataStyle).y
  const ctx = PaintBrush.getContext()

  return (
    sX < buffer ||
    sX > ctx.canvas.width - buffer ||
    sY < buffer ||
    sY > ctx.canvas.height - buffer
  )
}
