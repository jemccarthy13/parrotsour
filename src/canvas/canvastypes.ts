import { Point } from "classes/point"
import { FORMAT } from "../classes/supportedformats"
import { SensorType } from "classes/aircraft/datatrail/sensortype"
import { AircraftGroup } from "classes/groups/group"

export enum BlueInThe {
  NORTH,
  SOUTH,
  EAST,
  WEST,
}

export class FightAxis {
  public static isNS(orientation: BlueInThe): boolean {
    return orientation === BlueInThe.NORTH || orientation === BlueInThe.SOUTH
  }
}

export class CanvasOrient {
  height = 0
  width = 0
  orient = BlueInThe.EAST
}

export interface CanvasProps {
  orientation: CanvasOrient
  picType: string
  braaFirst: boolean
  dataStyle: SensorType
  showMeasurements: boolean
  isHardMode: boolean
  newPic: boolean
  animate: boolean
  resetCallback?: () => void
  animateCallback: () => void
}

export interface PictureCanvasState {
  bullseye: Point
  blueAir: AircraftGroup
  answer: PictureAnswer
  reDraw: PictureReDrawFunction
  animateCanvas?: ImageData
}

export interface CanvasDrawFunction {
  (context: CanvasRenderingContext2D): Promise<void>
}

export interface DrawCanvasProps extends CanvasProps {
  draw: CanvasDrawFunction
  bullseye: Point
  answer: PictureAnswer
}

export interface PictureCanvasProps extends CanvasProps {
  format: FORMAT
  setAnswer: { (answer: PictureAnswer): void }
  sliderSpeed: number
  desiredNumContacts: number
}

export type PictureAnswer = {
  pic: string
  groups: AircraftGroup[]
}

export interface PictureDrawFunction {
  (
    context: CanvasRenderingContext2D,
    props: PictureCanvasProps,
    state: PictureCanvasState,
    hasCaps: boolean,
    desiredNumContacts: number,
    start?: Point
  ): PictureAnswer
}

export interface PictureReDrawFunction {
  (forced?: boolean, start?: Point, hasCaps?: boolean): PictureAnswer
}
