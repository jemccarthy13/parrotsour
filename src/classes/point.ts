import { BlueInThe } from "../canvas/canvastypes"
import { BRAA } from "./braa"
import { PIXELS_TO_NM, toDegrees } from "../utils/psmath"

/**
 * An x,y Cartesian point that can calculate bearing and
 * range to another point.
 */
export class Point {
  x: number
  y: number

  static DEFAULT = new Point(0, 0)

  /**
   * Construct a new x,y point
   */
  constructor(x: number, y: number) {
    this.x = x
    this.y = y
  }

  /**
   * Get a bearing and range between this point and another point
   *
   * @param ptTo the "other" or "to" point
   * @returns the bearing and range between this (ptFrom) and ptTo
   */
  getBR(toPoint: Point): BRAA {
    const deltaX = this.x - toPoint.x
    const deltaY = this.y - toPoint.y

    // distance formula for range
    const rng = Math.floor(
      Math.sqrt(deltaX * deltaX + deltaY * deltaY) / PIXELS_TO_NM
    )

    // convert cartesian angle to heading
    let brg = Math.round(
      270 + toDegrees(Math.atan2(this.y - toPoint.y, this.x - toPoint.x))
    )
    if (brg > 360) {
      brg = brg - 360
    }

    return new BRAA(brg, rng)
  }

  /**
   * Get distance between points (straight along axis, i.e. not slant range)
   */
  public straightDistNM(toPoint: Point, orientation: BlueInThe): number {
    return Math.floor(
      Math.abs(
        orientation === BlueInThe.NORTH || orientation === BlueInThe.SOUTH
          ? this.y - toPoint.y
          : this.x - toPoint.x
      ) / PIXELS_TO_NM
    )
  }
}
