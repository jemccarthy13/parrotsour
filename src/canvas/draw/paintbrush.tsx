import { Braaseye } from "../../classes/braaseye"
import { AircraftGroup } from "../../classes/groups/group"
import { Point } from "../../classes/point"
import { PIXELS_TO_NM, randomNumber } from "../../utils/psmath"
import { PictureCanvasProps, PictureCanvasState } from "../canvastypes"
import { clampInContext } from "./drawutils"
import { formatAlt } from "./formatutils"

export class PaintBrush {
  private static ctx: CanvasRenderingContext2D

  constructor(ctx: CanvasRenderingContext2D) {
    PaintBrush.ctx = ctx
  }

  public static use(ctx: CanvasRenderingContext2D | undefined | null): void {
    if (ctx) PaintBrush.ctx = ctx
  }

  public static getContext(): CanvasRenderingContext2D {
    return PaintBrush.ctx
  }

  public static clearCanvas(): void {
    const ctx = PaintBrush.ctx
    ctx.clearRect(0, 0, ctx.canvas.width, ctx.canvas.height)
    ctx.fillStyle = "white"
    ctx.fillRect(0, 0, ctx.canvas.width, ctx.canvas.height)
  }

  /**
   * Draw a line on the drawing context, given the properties
   * @param startX start x of the line
   * @param startY start y of the line
   * @param endX end x of the line
   * @param endY end y of the line
   * @param color (optional) color of the line, defaults to black
   */
  public static drawLine(
    startX: number,
    startY: number,
    endX: number,
    endY: number,
    color = "black",
    forcedCtx?: CanvasRenderingContext2D
  ): void {
    let context = PaintBrush.ctx
    if (forcedCtx) context = forcedCtx
    context.lineWidth = 1
    context.strokeStyle = color
    context.beginPath()
    context.moveTo(startX, startY)
    context.lineTo(endX, endY)
    context.stroke()
    context.stroke()
  }

  /**
   * Draw text to the drawing context at the given position
   * @param text Text to draw
   * @param x X Position to draw at
   * @param y Y Position to draw at
   * @param size (optional) Size of the text to draw, defaults 12
   * @param color (optional) Color of the text, defaults to black
   */
  public static drawText(
    text: string,
    x: number,
    y: number,
    size = 12,
    color = "black"
  ): void {
    const ctx = PaintBrush.ctx
    ctx.lineWidth = 1
    ctx.fillStyle = color
    ctx.font = size + "px Arial"
    const pos = clampInContext(ctx, x, y)

    ctx.fillText(text, pos.x, pos.y)
    ctx.fillText(text, pos.x, pos.y)
  }

  /**
   * Draw a measurement (distance with a length number)
   * @param {Point} start position to start for line
   * @param {Point} end position to end for line
   * @param {number} distance how long the line should be
   * @param {boolean} showMeasurements Display the measurements
   */
  public static drawMeasurement(
    start: Point,
    end: Point,
    distance: number,
    showMeasurements: boolean
  ): void {
    if (showMeasurements) {
      PaintBrush.drawLine(start.x, start.y, end.x, end.y)
      PaintBrush.drawText(
        Math.floor(distance).toString(),
        (start.x + end.x) / 2,
        (start.y + end.y) / 2 - 3
      )
    }
  }

  /**
   * Draw altitudes next to a group with optional offset
   *
   * @param grpPos group's current position
   * @param alts altitudes to draw
   * @param offX (optional) x-axis offset from group
   * @param offY (optional) y-axis offset from group
   */
  public static drawAltitudes(
    grpPos: Point,
    alts: number[],
    offX?: number,
    offY?: number
  ): void {
    const offsetX = offX || 0
    const offsetY = offY || 0
    const formattedAlts: string[] = alts.map((a: number) => {
      return formatAlt(a)
    })
    PaintBrush.drawText(
      formattedAlts.join(","),
      grpPos.x + 25 + offsetX,
      grpPos.y - 11 + offsetY,
      11,
      "#ff8c00"
    )
  }

  public static drawBullseye(bull?: Point, color?: string): Point {
    const context = PaintBrush.ctx
    color = color || "black"
    context.lineWidth = 1
    context.fillStyle = color
    context.strokeStyle = color

    const centerPointX = bull
      ? bull.x
      : randomNumber(context.canvas.width * 0.33, context.canvas.width * 0.66)
    const centerPointY = bull
      ? bull.y
      : randomNumber(context.canvas.height * 0.33, context.canvas.height * 0.66)

    context.beginPath()
    context.arc(
      centerPointX,
      centerPointY,
      PIXELS_TO_NM / 2,
      0,
      2 * Math.PI,
      true
    )
    context.stroke()
    context.fill()

    context.moveTo(centerPointX, centerPointY + PIXELS_TO_NM * 2)
    context.lineTo(centerPointX, centerPointY - PIXELS_TO_NM * 2)
    context.stroke()

    context.moveTo(centerPointX + PIXELS_TO_NM * 2, centerPointY)
    context.lineTo(centerPointX - PIXELS_TO_NM * 2, centerPointY)
    context.stroke()

    return new Point(centerPointX, centerPointY)
  }

  public static drawFullInfo(
    state: PictureCanvasState,
    props: PictureCanvasProps,
    groups: AircraftGroup[]
  ): void {
    for (let y = 0; y < groups.length; y++) {
      const grpPos = groups[y].getCenterOfMass(props.dataStyle)
      if (props.showMeasurements) {
        new Braaseye(
          grpPos,
          state.blueAir.getCenterOfMass(props.dataStyle),
          state.bullseye
        ).draw(true, props.braaFirst)
      }
      PaintBrush.drawAltitudes(grpPos, groups[y].getAltitudes())
    }
  }
}
