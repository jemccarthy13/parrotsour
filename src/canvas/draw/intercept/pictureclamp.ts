import { BlueInThe, FightAxis } from "../../../canvas/canvastypes"
import { SensorType } from "../../../classes/aircraft/datatrail/sensortype"
import { AircraftGroup } from "../../../classes/groups/group"
import { Point } from "../../../classes/point"
import { PIXELS_TO_NM, randomNumber } from "../../../utils/psmath"
import { PaintBrush } from "../paintbrush"

export type Bounds = {
  lowX: number
  hiX: number
  lowY: number
  hiY: number
}

export interface PictureInfo {
  deep?: number
  wide?: number
  start?: Point
}

//
// TODO -- Comment this file so people know what's happening here.
//
//
// TODO -- JEST -- "how to test non-exported functions"
// so I can take the export off of all the helper functions
// necessary to export for now for testing/coverage
//
export const _howFarOut = (x: number, min: number, max: number): number => {
  if (x < min) {
    return min - x
  } else if (x > max) {
    return -(x - max)
  }
  return 0
}

export const _clampPictureInContext = (
  pInfo: PictureInfo,
  orientation: BlueInThe
): Point => {
  const ctx = PaintBrush.getContext()
  if (!pInfo.start) {
    console.warn(
      "Without a starting point to clamp, _clampPictureInContext will generate a random point."
    )
    return new Point(
      randomNumber(1, ctx.canvas.width),
      randomNumber(1, ctx.canvas.height)
    )
  }

  const isNS = FightAxis.isNS(orientation)

  const width = pInfo.wide || 7 * PIXELS_TO_NM
  const depth = pInfo.deep || 7 * PIXELS_TO_NM

  const minValWithBuffer = 1 + 7 * PIXELS_TO_NM
  const maxXWithBuffer = ctx.canvas.width - 1 - (isNS ? width : depth)
  const maxYWithBuffer =
    ctx.canvas.height - 1 - 7 * PIXELS_TO_NM - (isNS ? depth : width)

  pInfo.start.x += _howFarOut(pInfo.start.x, minValWithBuffer, maxXWithBuffer)
  pInfo.start.y += _howFarOut(pInfo.start.y, minValWithBuffer, maxYWithBuffer)
  return pInfo.start
}

export const getRestrictedStartPos = (
  blueAir: AircraftGroup,
  orientation: BlueInThe,
  dataStyle: SensorType,
  minNMFromBlue: number,
  maxNMFromBlue: number,
  pInfo?: PictureInfo
): Point => {
  const ctx = PaintBrush.getContext()
  const blueLoc = blueAir.getCenterOfMass(dataStyle)

  let limitLine = blueLoc.x
  let canvasSize = ctx.canvas.width
  if (FightAxis.isNS(orientation)) {
    limitLine = blueLoc.y
    canvasSize = ctx.canvas.height
  }

  let lBound = (limitLine + maxNMFromBlue * PIXELS_TO_NM) / canvasSize
  let uBound = (limitLine + minNMFromBlue * PIXELS_TO_NM) / canvasSize
  if (orientation === BlueInThe.SOUTH || orientation === BlueInThe.EAST) {
    lBound = (limitLine - maxNMFromBlue * PIXELS_TO_NM) / canvasSize
    uBound = (limitLine - minNMFromBlue * PIXELS_TO_NM) / canvasSize
  }

  const isNS =
    orientation === BlueInThe.NORTH || orientation === BlueInThe.SOUTH

  let mults = { lowX: lBound, hiX: uBound, lowY: 0.2, hiY: 0.8 }
  if (isNS) {
    mults = { lowX: 0.2, hiX: 0.8, lowY: lBound, hiY: uBound }
  }

  if (pInfo?.start) {
    console.warn(
      "Providing a start point to getRestrictedStartPos will ignore restrictions " +
        "and result in no change to Point."
    )
  }

  const startY: number =
    (pInfo?.start && pInfo.start.y) ||
    randomNumber(ctx.canvas.height * mults.lowY, ctx.canvas.height * mults.hiY)
  const startX: number =
    (pInfo?.start && pInfo.start.x) ||
    randomNumber(ctx.canvas.width * mults.lowX, ctx.canvas.width * mults.hiX)

  pInfo = {
    deep: pInfo?.deep || 7 * PIXELS_TO_NM,
    wide: pInfo?.wide || 7 * PIXELS_TO_NM,
    start: new Point(startX, startY),
  }

  const point = _clampPictureInContext(pInfo, orientation)

  return point
}

export const getStartPos = (
  blueAir: AircraftGroup,
  orientation: BlueInThe,
  dataStyle: SensorType,
  pInfo?: PictureInfo
): Point => {
  const ctx = PaintBrush.getContext()
  let canvasSize = ctx.canvas.width
  if (orientation === BlueInThe.NORTH || orientation === BlueInThe.SOUTH) {
    canvasSize = ctx.canvas.height
  }
  const maxMiles = canvasSize / PIXELS_TO_NM
  return getRestrictedStartPos(
    blueAir,
    orientation,
    dataStyle,
    45,
    maxMiles,
    pInfo
  )
}
