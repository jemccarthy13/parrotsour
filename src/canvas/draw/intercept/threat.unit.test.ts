import { SensorType } from "../../../classes/aircraft/datatrail/sensortype"
import { AircraftGroup } from "../../../classes/groups/group"
import { Point } from "../../../classes/point"
import { PIXELS_TO_NM } from "../../../utils/psmath"
import { BlueInThe, PictureCanvasState } from "../../canvastypes"
import { PaintBrush } from "../paintbrush"
import { testProps } from "./mockutils.unit.test"
import { PictureInfo } from "./pictureclamp"

import DrawThreat from "./threat"

// eslint-disable-next-line @typescript-eslint/no-var-requires
const canvasSerializer = require("jest-canvas-snapshot-serializer")
expect.addSnapshotSerializer(canvasSerializer)

describe("DrawThreat", () => {
  let dThreat: DrawThreat
  const canvas = document.createElement("canvas")
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const ctx = canvas.getContext("2d")!
  canvas.width = 800
  canvas.height = 500

  let testState: PictureCanvasState = {
    bullseye: new Point(400, 400),
    blueAir: new AircraftGroup({ sx: 600, sy: 200, hdg: 270, nContacts: 4 }),
    answer: { pic: "3 grp ladder", groups: [] },
    reDraw: jest.fn(),
  }

  beforeEach(() => {
    jest.restoreAllMocks()

    PaintBrush.use(ctx)

    dThreat = new DrawThreat()

    testState = {
      bullseye: new Point(400, 400),
      blueAir: new AircraftGroup({ sx: 600, sy: 200, hdg: 270, nContacts: 4 }),
      answer: { pic: "3 grp ladder", groups: [] },
      reDraw: jest.fn(),
    }

    dThreat.initialize(testProps, testState)
  })

  it("drawinfo_threat", () => {
    expect(true).toEqual(true)

    const p = {
      dataTrailType: SensorType.ARROW,
      sx: 500,
      sy: 200,
      nContacts: 1,
      hdg: 90,
      alts: [20],
    }
    dThreat.groups = [new AircraftGroup({ ...p })]

    dThreat.drawInfo()

    expect(canvas).toMatchSnapshot()
  })

  it("hot_threat_answer", () => {
    expect(true).toEqual(true)

    const p = {
      dataTrailType: SensorType.ARROW,
      sx: 500,
      sy: 200,
      nContacts: 1,
      hdg: 90,
      alts: [20],
    }
    dThreat.groups = [new AircraftGroup({ ...p })]

    dThreat.drawInfo()
    expect(dThreat.getAnswer()).toEqual(
      "[FTR C/S], THREAT GROUP BRAA " + "257/13 20k HOT HOSTILE"
    )
  })

  it("flank_threat_answer", () => {
    expect(true).toEqual(true)

    const p = {
      dataTrailType: SensorType.ARROW,
      sx: 500,
      sy: 200,
      nContacts: 1,
      hdg: 110,
      alts: [20],
    }
    dThreat.groups = [new AircraftGroup({ ...p })]

    dThreat.drawInfo()
    expect(dThreat.getAnswer()).toEqual(
      "[FTR C/S], THREAT GROUP BRAA " + "250/14 20k FLANK SOUTHEAST HOSTILE"
    )
  })

  it("beam_n_threat_answer", () => {
    expect(true).toEqual(true)

    const p = {
      dataTrailType: SensorType.ARROW,
      sx: 500,
      sy: 200,
      nContacts: 1,
      hdg: 10,
      alts: [20],
    }
    dThreat.groups = [new AircraftGroup({ ...p })]

    dThreat.drawInfo()
    expect(dThreat.getAnswer()).toEqual(
      "[FTR C/S], THREAT GROUP BRAA " + "279/18 20k BEAM NORTH HOSTILE"
    )
  })

  it("beam_s_threat_answer", () => {
    expect(true).toEqual(true)

    const p = {
      dataTrailType: SensorType.ARROW,
      sx: 500,
      sy: 200,
      nContacts: 1,
      hdg: 170,
      alts: [20],
    }
    dThreat.groups = [new AircraftGroup({ ...p })]

    dThreat.drawInfo()
    expect(dThreat.getAnswer()).toEqual(
      "[FTR C/S], THREAT GROUP BRAA " + "244/20 20k BEAM SOUTH HOSTILE"
    )
  })

  it("drag_threat_answer", () => {
    expect(true).toEqual(true)

    const p = {
      dataTrailType: SensorType.ARROW,
      sx: 500,
      sy: 200,
      nContacts: 1,
      hdg: 270,
      alts: [20],
    }
    dThreat.groups = [new AircraftGroup({ ...p })]

    dThreat.drawInfo()
    expect(dThreat.getAnswer()).toEqual(
      "[FTR C/S], THREAT GROUP BRAA " + "263/25 20k DRAG WEST HOSTILE"
    )
  })

  it("tests_benign_functions", () => {
    expect(dThreat.formatWeighted()).toEqual("")
    expect(dThreat.formatDimensions()).toEqual("")

    dThreat.create()
    dThreat.groups = dThreat.createGroups(new Point(100, 100), [1])
    expect(dThreat.getNumGroups()).toEqual(1)
    expect(dThreat.groups.length).toEqual(1)

    expect(dThreat.numGroupsToCreate).toEqual(0)
    dThreat.chooseNumGroups()
    expect(dThreat.numGroupsToCreate).toEqual(1)
  })

  it("tests_getPictureInfo_with_start", () => {
    const startPt = new Point(100, 100)
    const pInfo: PictureInfo = dThreat.getPictureInfo(startPt)
    expect(pInfo.start).toEqual(startPt)
    expect(pInfo.deep).toEqual(5 * PIXELS_TO_NM)
    expect(pInfo.wide).toEqual(5 * PIXELS_TO_NM)
  })

  it("tests_getPictureInfo_random_start", () => {
    const pInfo: PictureInfo = dThreat.getPictureInfo()

    const bPos = testState.blueAir.getCenterOfMass(testProps.dataStyle)

    const start = pInfo.start

    expect(start).toBeDefined()
    if (start) {
      const startBraa = bPos.getBR(start)
      expect(startBraa.range).toBeLessThan(30)
    }
  })

  it("tests_getPictureInfo_random_start_NS", () => {
    const newProps = { ...testProps }
    newProps.orientation.orient = BlueInThe.NORTH
    dThreat.initialize(newProps, testState)

    const pInfo: PictureInfo = dThreat.getPictureInfo()
    const bPos = testState.blueAir.getCenterOfMass(testProps.dataStyle)
    const start = pInfo.start

    expect(start).toBeDefined()
    if (start) {
      const startBraa = bPos.getBR(start)
      expect(startBraa.range).toBeLessThan(30)
    }
  })
  // TODO -- getPictureInfo
})
