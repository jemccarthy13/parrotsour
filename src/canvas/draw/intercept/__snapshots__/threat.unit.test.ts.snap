// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawThreat drawinfo_threat 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAyAAAAH0CAYAAADFQEl4AAAABmJLR0QA/wD/AP+gvaeTAAAKUUlEQVR4nO3dz6tndR3H8edVGSaDtEEyEbN1iAjiRmbTxgKJXAuziRmSitCcAmGqCYkCr2Zpm9BVkItWJQVBi1b2BwSSzCIkCnNM/DHN5Nzxflqcc+/93q/33iFw8M7M4wFfvue8zudzfiwu3Defz+d8CwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAgMvGykd9AwAAVWO1o9XPqo9V57q2r6480i/HEx1rpWerA610ttH3Vo735G75R/oQwEVd81HfAADA7Oet92Rn+mT1m97vF1Wt9IPqt93Qpxv9ufr2njkAAMD/Y5zsprHaetV4sgtjtdurxmp3j9XO75UD+5sREABg//l4x1rpr1WNru1M/6rqQKeq6/bMgX1NAQIA7Cvjie6svlV9YSPqUAeqenPb/y675cA+5o8VANg3xk+6q3qxa/ryyqP9vaqVLjT6TFU39Nmap1rtlgP7mgIEANgXxlOdaL0/daD7Vx7tpYVD/+hCPxpPdaj1fli9eZEc2MfMlQQA9ofR9xtd11p/GatTtHK8ldb7aSutNvp3db7RN6p2zQEAAAAAAAAAAAAAALiExtEaZ2uM+fvInB+es4VPfTDbyDf7rM3bn6/x/tK1Hqjx7tzvTI3jl/75AACAfWS8V+PxGjfWeKHGuTl/rsYrF+n7qxq/X9h/rsapGj+ucX57cVJzgfN8jVtqnNwqVgAAgKvQuKnG+rx9qsYze7T9Yo23axxYyE7NRcjrNb70wQJks91tc7u3P7x7BwAALjPjsRovz9trNd6ap0u9U+P+pbav1fjOUrY2TcPa3N+hABm3TkXOGDWe/nDvHwAAuEyMO2ucnkYntuUHazw4rd3YzB6YplNta7ew/mMz22UEpGo8O03TAgAArjLjrhqv1rh3l+MHt6ZmVY2Xavxhqc28/mNbtlcBcvPexwEAgCvQODFPtbpjKT9f46G5+Dha442FY2dqPLLUfl7/sS1bXoT+3tRvXF9jdVorAgAAXEXG2i6v2z06vRFrrE/Fx7hnoc96jc/tcJ7DS9lyAXKkrVf+nt5+TgAAAAAAAAAAAAAAAPZwrHqvGtV/qkfn/IHq3Tk/Ux2f86PV2Tk/Wx1ZONfham3+Hkufqoercwt9v3YpHggAANi//ln9urq5+mP12pyfrZ6vbqlONhUWNRUrj1c3Vi80FRQbnqtOzd+v7HCtC9VXqgNNhc/aDm0AAIAr2IXq9nn77mr5xwFvayoo3t6h703Vwu+DbBYfp6pndmh/vq0C5MgO1wIAAK5wozo4b3+i7QXFrfP+qJ7eoe9j1csL+xvTr9aqt+Z+71T3z8e/2fZpWV//UJ4AAAC4bKw3FR41Tata36HNs31wtOLO6nTTCEltFR6LDlYPNq0lqWktyYn5eo+386gKAABwBTtfbfwK+l3Vf3doc3NbC8k32r1a3buQbUy9WnawraJmvbp+3v7U0jkBAICrwN+qF6tD1e+aFqXXtNj8kaaCYbV6fc5PNE2vumP7aTbXf9RU1DzUVHwcrd6Y8zebpl0dqr6bERAAALjqPNy0EH00FR3H5vxIW6/bPV3dM+dr7fyK3Y31HzUVHeeaRjzeWOh7X1PRMebv+y7FAwEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAFx2/gcY4sz2N+bjBAAAAABJRU5ErkJggg=="
  height="500"
  width="800"
/>
`;
